src/states/establishing_node.rs

```diff
@@ -0,0 +1,463 @@
+    // Can also just be a single target once node-ageing makes Offline votes Opaque which should
+    // remove invalid test failures for unaccumulated parsec::Remove blocks.
+    fn send_parsec_poke(&mut self) {
+        let version = *self.gen_pfx_info.first_info.version();
+        let recipients = self
+            .gen_pfx_info
+            .latest_info
+            .members()
+            .iter()
+            .cloned()
+            .collect_vec();
+
+        for recipient in recipients {
+            self.send_message(
+                &recipient,
+                Message::Direct(DirectMessage::ParsecPoke(version)),
+            );
+        }
+    }
+}
+
+#[cfg(feature = "mock")]
+impl EstablishingNode {
+    pub fn get_timed_out_tokens(&mut self) -> Vec<u64> {
+        self.timer.get_timed_out_tokens()
+    }
+
+    pub fn has_unconsensused_observations(&self) -> bool {
+        self.parsec.has_unconsensused_observations()
+    }
+}
+
+impl Base for EstablishingNode {
+    fn crust_service(&self) -> &Service {
+        &self.crust_service
+    }
+
+    fn full_id(&self) -> &FullId {
+        &self.full_id
+    }
+
+    fn in_authority(&self, auth: &Authority<XorName>) -> bool {
+        if let Authority::Client { ref client_id, .. } = *auth {
+            client_id == self.full_id.public_id()
+        } else {
+            false
+        }
+    }
+
+    fn min_section_size(&self) -> usize {
+        self.chain.min_sec_size()
+    }
+
+    fn handle_timeout(&mut self, token: u64, _: &mut EventBox) -> Transition {
+        if self.poke_timer_token == token {
+            self.send_parsec_poke();
+            self.poke_timer_token = self.timer.schedule(POKE_TIMEOUT);
+        } else {
+            self.resend_unacknowledged_timed_out_msgs(token);
+        }
+
+        Transition::Stay
+    }
+
+    fn handle_connect_success(&mut self, pub_id: PublicId, outbox: &mut EventBox) -> Transition {
+        Relocated::handle_connect_success(self, pub_id, outbox)
+    }
+
+    fn handle_connect_failure(&mut self, pub_id: PublicId, _: &mut EventBox) -> Transition {
+        RelocatedNotEstablished::handle_connect_failure(self, pub_id)
+    }
+
+    fn handle_direct_message(
+        &mut self,
+        msg: DirectMessage,
+        pub_id: PublicId,
+        outbox: &mut EventBox,
+    ) -> Result<Transition, RoutingError> {
+        self.check_direct_message_sender(&msg, &pub_id)?;
+
+        use crate::messages::DirectMessage::*;
+        match msg {
+            ParsecRequest(version, par_request) => {
+                self.handle_parsec_request(version, par_request, pub_id, outbox)
+            }
+            ParsecResponse(version, par_response) => {
+                self.handle_parsec_response(version, par_response, pub_id, outbox)
+            }
+            _ => {
+                debug!("{} Unhandled direct message: {:?}", self, msg);
+                Ok(Transition::Stay)
+            }
+        }
+    }
+
+    fn handle_hop_message(
```

Comment:
Looking at node.rs handle_hop_message it is not clear to me how we have that implementation:

```rust
    fn handle_hop_message(
        &mut self,
        hop_msg: HopMessage,
        pub_id: PublicId,
        _: &mut EventBox,
    ) -> Result<Transition, RoutingError> {
        hop_msg.verify(pub_id.signing_public_key())?;
        let mut client_ip = None;
        let mut hop_name_result = match self.peer_mgr.get_peer(&pub_id).map(Peer::state) {
            Some(&PeerState::Bootstrapper { .. }) => {
                warn!(
                    "{} Hop message received from bootstrapper {:?}, disconnecting.",
                    self, pub_id
                );
                Err(RoutingError::InvalidStateForOperation)
            }
            Some(&PeerState::Client { ip, .. }) => {
                client_ip = Some(ip);
                Ok(*self.name())
            }
            Some(&PeerState::JoiningNode) => Ok(*self.name()),
            Some(&PeerState::Candidate) | Some(&PeerState::Proxy) | Some(&PeerState::Routing) => {
                Ok(*pub_id.name())
            }
            Some(&PeerState::ConnectionInfoPreparing { .. })
            | Some(&PeerState::ConnectionInfoReady(_))
            | Some(&PeerState::CrustConnecting)
            | Some(&PeerState::Connected)
            | None => {
                if self.dropped_clients.contains_key(&pub_id) {
                    debug!(
                        "{} Ignoring {:?} from recently-disconnected client {:?}.",
                        self, hop_msg, pub_id
                    );
                    return Ok(Transition::Stay);
                } else {
                    Ok(*self.name())
                    // FIXME - confirm we can return with an error here by running soak tests
                    // debug!("{} Invalid sender {} of {:?}", self, pub_id, hop_msg);
                    // return Err(RoutingError::InvalidSource);
                }
            }
        };

        if let Some(ip) = client_ip {
            match self.check_valid_client_message(&ip, hop_msg.content.routing_message()) {
                Ok(added_bytes) => {
                    self.proxy_load_amount += added_bytes;
                    self.peer_mgr.add_client_traffic(&pub_id, added_bytes);
                }
                Err(e) => hop_name_result = Err(e),
            }
        }

        match hop_name_result {
            Ok(hop_name) => {
                let HopMessage {
                    content,
                    route,
                    sent_to,
                    ..
                } = hop_msg;
                self.handle_signed_message(content, route, hop_name, &sent_to)
                    .map(|()| Transition::Stay)
            }
            Err(RoutingError::ExceedsRateLimit(hash)) => {
                trace!(
                    "{} Temporarily can't proxy messages from client {:?} (rate-limit hit).",
                    self,
                    pub_id
                );
                self.send_direct_message(
                    pub_id,
                    DirectMessage::ProxyRateLimitExceeded {
                        ack: Ack::compute(hop_msg.content.routing_message())?,
                    },
                );
                Err(RoutingError::ExceedsRateLimit(hash))
            }
            Err(error) => {
                self.ban_and_disconnect_peer(&pub_id);
                Err(error)
            }
        }
    }

```

Reply:
This isn't resolved yet.

Reply:
After spending some time on this I'm now about 90% convinced this is actually correct. Although it's a logic change I believe the way this was handled before was actually somehow buggy (although in a harmless way) because until a node is established, it isn't supposed to handle these things. That's why I'd prefer to keep things as they are now. Worst case, this will pop up in the soak testing at which point I'll change it back. I'm going to at least mention this in the PR description.